RerouteIT::testReroutingToErrorView is flaky

`RerouteIT.testReroutingToErrorView` sometimes fails.

https://bender.vaadin.com/viewLog.html?buildId=212186&tab=buildResultsDiv&buildTypeId=Flow_Flow24_Nightly_FlowFelixJettyIntegrationTests24

The screenshot shows a loading indicator so it might be a timing issue.

```
Caused by: java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertTrue(Assert.java:53)
	at com.vaadin.flow.uitest.ui.RerouteIT.testReroutingToErrorView(RerouteIT.java:36)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

--- a/flow-tests/test-root-context/src/main/java/com/vaadin/flow/osgi/OSGiMarker.java
+++ b/flow-tests/test-root-context/src/main/java/com/vaadin/flow/osgi/OSGiMarker.java
@@ -1,0 +1,25 @@
+
+/*
+ * Copyright 2000-2020 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.osgi;
+
+/**
+ * This interface is a marker of OSGi.
+ *
+ */
+public interface OSGiMarker {
+
+}

--- a/flow-tests/test-root-context/src/main/java/com/vaadin/flow/osgi/OSGiMarkerImpl.java
+++ b/flow-tests/test-root-context/src/main/java/com/vaadin/flow/osgi/OSGiMarkerImpl.java
@@ -1,0 +1,14 @@
+package com.vaadin.flow.osgi;
+
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * 
+ * Implementation of {@link OSGiMarker} : presence of {@link OSGiMarker} service
+ * in {@link Lookup} means the app is executed in OSGi env because standard
+ * Lookup doesn't know anything about {@link OSGiMarker}.
+ */
+@Component(immediate = true)
+public class OSGiMarkerImpl implements OSGiMarker {
+
+}

--- a/flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/routing/PushLayout.java
+++ b/flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/routing/PushLayout.java
@@ -15,14 +15,19 @@
  */
 package com.vaadin.flow.uitest.ui.routing;
 
+import com.vaadin.flow.component.PushConfiguration;
+import com.vaadin.flow.component.UI;
 import com.vaadin.flow.component.html.Div;
-import com.vaadin.flow.component.page.Push;
+import com.vaadin.flow.di.Lookup;
+import com.vaadin.flow.osgi.OSGiMarker;
 import com.vaadin.flow.router.BeforeEnterEvent;
 import com.vaadin.flow.router.BeforeEnterObserver;
 import com.vaadin.flow.router.Route;
 import com.vaadin.flow.router.RouterLayout;
+import com.vaadin.flow.server.VaadinService;
+import com.vaadin.flow.shared.communication.PushMode;
+import com.vaadin.flow.shared.ui.Transport;
 
-@Push
 @Route("com.vaadin.flow.uitest.ui.PushLayout")
 public class PushLayout extends Div
         implements RouterLayout, BeforeEnterObserver {
@@ -31,6 +36,14 @@ public class PushLayout extends Div
 
     public PushLayout() {
         setId("push-layout");
+        Lookup lookup = VaadinService.getCurrent().getContext()
+                .getAttribute(Lookup.class);
+        if (lookup.lookup(OSGiMarker.class) == null) {
+            PushConfiguration pushConfiguration = UI.getCurrent()
+                    .getPushConfiguration();
+            pushConfiguration.setPushMode(PushMode.AUTOMATIC);
+            pushConfiguration.setTransport(Transport.WEBSOCKET_XHR);
+        }
     }
 
     @Override

--- a/flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/routing/PushLayout.java
+++ b/flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/routing/PushLayout.java
@@ -18,38 +18,44 @@ package com.vaadin.flow.uitest.ui.routing;
 import com.vaadin.flow.component.PushConfiguration;
 import com.vaadin.flow.component.UI;
 import com.vaadin.flow.component.html.Div;
+import com.vaadin.flow.component.page.Push;
 import com.vaadin.flow.di.Lookup;
 import com.vaadin.flow.osgi.OSGiMarker;
+import com.vaadin.flow.router.AfterNavigationEvent;
+import com.vaadin.flow.router.AfterNavigationObserver;
 import com.vaadin.flow.router.BeforeEnterEvent;
 import com.vaadin.flow.router.BeforeEnterObserver;
 import com.vaadin.flow.router.Route;
 import com.vaadin.flow.router.RouterLayout;
 import com.vaadin.flow.server.VaadinService;
 import com.vaadin.flow.shared.communication.PushMode;
-import com.vaadin.flow.shared.ui.Transport;
 
+@Push
 @Route("com.vaadin.flow.uitest.ui.PushLayout")
 public class PushLayout extends Div
-        implements RouterLayout, BeforeEnterObserver {
+        implements RouterLayout, BeforeEnterObserver, AfterNavigationObserver {
 
     public static String FORWARD_PATH = "forward-no-route";
 
     public PushLayout() {
         setId("push-layout");
-        Lookup lookup = VaadinService.getCurrent().getContext()
-                .getAttribute(Lookup.class);
-        if (lookup.lookup(OSGiMarker.class) == null) {
-            PushConfiguration pushConfiguration = UI.getCurrent()
-                    .getPushConfiguration();
-            pushConfiguration.setPushMode(PushMode.AUTOMATIC);
-            pushConfiguration.setTransport(Transport.WEBSOCKET_XHR);
-        }
     }
 
     @Override
     public void beforeEnter(BeforeEnterEvent event) {
         if (FORWARD_PATH.equals(event.getLocation().getPath())) {
             event.forwardTo(ForwardPage.class);
+        }
+    }
+
+    @Override
+    public void afterNavigation(AfterNavigationEvent event) {
+        Lookup lookup = VaadinService.getCurrent().getContext()
+                .getAttribute(Lookup.class);
+        if (lookup.lookup(OSGiMarker.class) != null) {
+            PushConfiguration pushConfiguration = UI.getCurrent()
+                    .getPushConfiguration();
+            pushConfiguration.setPushMode(PushMode.DISABLED);
         }
     }
 

